src/mono/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,140 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Runtime
+{
+    internal struct Ephemeron
+    {
+        public object? Key;
+        public object? Value;
+    }
+
+    //
+    // Instead of dependent handles, mono uses arrays of Ephemeron objects.
+    //
+    public struct DependentHandle : IDisposable
+    {
+        private Ephemeron[]? _data;
+
+        public DependentHandle(object? target, object? dependent)
+        {
+            _data = new Ephemeron[1];
+            _data[0].Key = target;
+            _data[0].Value = dependent;
+            GC.register_ephemeron_array(_data);
+        }
+
+        public bool IsAllocated => _data is not null;
+
+        public object? Target
+        {
+            get => UnsafeGetTarget();
+            set => UnsafeSetTarget(value);
+        }
+
+        public object? Dependent
+        {
+            get => UnsafeGetDependent();
+            set => UnsafeSetDependent(value);
+        }
+
+        public (object? Target, object? Dependent) GetTargetAndDependent()
+        {
+            object? target = UnsafeGetTargetAndDependent(out object? dependent);
+
+            return (target, dependent);
+        }
+
+        internal object? UnsafeGetTarget()
+        {
+            Ephemeron[]? data = _data;
+
+            if (data is null)
+            {
+                ThrowHelper.ThrowInvalidOperationException();
+
+                return default;
+            }
+
+            object? key = data[0].Key;
+
+            return key != GC.EPHEMERON_TOMBSTONE ? key : null;
+        }
+
+        internal object? UnsafeGetDependent()
+        {
+            Ephemeron[]? data = _data;
+
+            if (data is null)
+            {
+                ThrowHelper.ThrowInvalidOperationException();
+
+                return default;
+            }
+
+            Ephemeron e = data[0];
```

Comment:
@vargaz `Ephemeron` type is treated in a special way by the Mono runtime. Is it ok to create a local of `Ephemeron` type like this?

Reply:
Mono treats those as having no ref fields, but using it like this shouldn't be a problem I think.

Reply:
Hey @vargaz, I'm a bit confused by that and would love to understand how this works and why that is safe a bit better, would you mind sharing some more details? Specifically, when you said that "Mono treats those [`Ephemeral` instances] as having no ref fields", that makes me wonder: in that case wouldn't it be possible that if a compacting GC was triggered right after loading that local and before reading its fields, and in case that GC moved the referenced objects, those references stored in the local could become invalid? As in, if the runtime treats the type as having no ref fields that would make me think the GC wouldn't be properly tracking the object references in its fields either and updating them as needed? Am I reading this wrong or maybe I'm just missing something here? Thanks! 😄

Reply:
The current mono gc does a conservative stack scan so it will pick up those references. The mono gc special cases arrays of Ephemeron, which is a historical thing because those were used to implement ConditionalWeakMap. The old code used data[0].Key, which I think also creates a Ephemeron local on the stack, so the two should be equivalent.

Reply:
Oh I see, makes sense!
Thank you for taking the time to explain that! 🙂